src/Form/AuthenticationForm.php

```diff
@@ -456,77 +490,85 @@ protected function createSuggestion(\Exception $exception, KeyInterface $key): M
-
-      $keys['key_token_type'] = get_class($key_token->getKeyType());
-      $keys['key_token_provider'] = get_class($key_token->getKeyProvider());
     }
 
-    $exception_text = (string) $exception;
-    $exception_text = preg_replace('/(.*refresh_token=)([^\&\r\n]+)(.*)/', '$1***refresh-token***$3', $exception_text);
-    $exception_text = preg_replace('/(.*mfa_token=)([^\&\r\n]+)(.*)/', '$1***mfa-token***$3', $exception_text);
-    $exception_text = preg_replace('/(.*password=)([^\&\r\n]+)(.*)/', '$1***password***$3', $exception_text);
-    $exception_text = preg_replace('/(Authorization: (Basic|Bearer) ).*/', '$1***credentials***', $exception_text);
-
-    $text = json_encode($credentials, JSON_PRETTY_PRINT) .
-      PHP_EOL .
-      json_encode($keys, JSON_PRETTY_PRINT) .
-      PHP_EOL .
-      json_encode($this->config('apigee_edge.client')->get(), JSON_PRETTY_PRINT) .
-      PHP_EOL .
+    // Sanitize exception text.
+    $exception_text = preg_replace([
+      '/(.*refresh_token=)([^\&\r\n]+)(.*)/',
+      '/(.*mfa_token=)([^\&\r\n]+)(.*)/',
+      '/(.*password=)([^\&\r\n]+)(.*)/',
+      '/(Authorization: (Basic|Bearer) ).*/',
+    ], [
+      '$1***refresh-token***$3',
+      '$1***mfa-token***$3',
+      '$1***password***$3',
+      '$1***credentials***',
+    ], (string) $exception);
+
+    return json_encode($credentials, JSON_PRETTY_PRINT) . PHP_EOL .
+      json_encode($keys, JSON_PRETTY_PRINT) . PHP_EOL .
+      json_encode($this->config('apigee_edge.client')->get(), JSON_PRETTY_PRINT) . PHP_EOL .
       $exception_text;
-
-    return $text;
   }
 
   /**
    * Pass form array to the AJAX callback.
    *
-   * @param array $form
+   * @param array &$form
    *   An associative array containing the structure of the form.
    * @param \Drupal\Core\Form\FormStateInterface $form_state
    *   Form state object.
    *
-   * @return array
+   * @return \Drupal\Core\Ajax\AjaxResponse
    *   The AJAX response.
+   *
+   * @throws \Exception
    */
-  public function ajaxCallback(array $form, FormStateInterface $form_state): array {
-    return $form;
+  public function ajaxCallback(array &$form, FormStateInterface $form_state): AjaxResponse {
+    $response = new AjaxResponse();
+    // Get any status messages so they can be rendered in the placeholder.
+    $status_messages = $this->renderer->render(StatusMessages::renderMessages());
+    // Clear any existing messages from the initial page load.
+    $response->addCommand(new ReplaceCommand('div.messages', ''));
+    $response->addCommand(new ReplaceCommand('#apigee-edge-auth-form-messages', "<div id=\"apigee-edge-auth-form-messages\">{$status_messages}</div>"));
+
+    // Only render the debug element if it's been enabled by validation.
+    if ($form['debug']['#access']) {
+      // Add a prefix to the debug section to replace the wrapper.
```

Comment:
But this "wrapper" could have been set on the form element even if it is not visible by default, isn't it?
https://github.com/apigee/apigee-edge-drupal/pull/103/files#diff-c1ab508d8adb4da185ab3867ced29eb6R210

Reply:
This is true. I as trying to solve for a case before I started returning an ajax response instead of a form element.